### PR TITLE
Fix trajectory execution manager tests

### PR DESCRIPTION
**Pull request by cambel**
I converted the trajectory execution manager tests to Gtest with the panda robot, as part of the #3156.
A fix is added to the `execute()`: Skip execution if there is no trajectory pushed.

### CODE REVIEW

File: moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp [hunk 1]
```diff
@@ -1,107 +1,103 @@
...
- *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
- *  POSSIBILITY OF SUCH DAMAGE.
- *********************************************************************/
-
-/* Author: Ioan Sucan */
-
-#include <moveit/trajectory_execution_manager/trajectory_execution_manager.h>
-#include <moveit/robot_model_loader/robot_model_loader.h>
-#include <moveit/planning_scene_monitor/planning_scene_monitor.h>
+/* Author: Cristian C. Beltran
```
Comment (rhaschke) [inline moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp:1]: Please keep a copyright notice.

File: moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp [hunk 2]
```diff
@@ -1,107 +1,103 @@
...
+  moveit_msgs::RobotTrajectory traj2;
+};
 
-  std::cout << "4:\n";
-  if (!tem.ensureActiveControllersForJoints(std::vector<std::string>(1, "lj1")))
-    ROS_ERROR("Fail!");
+TEST_F(MoveItCppTest, EnsureActiveControllersForJointsTest)
+{
+  ASSERT_TRUE(
+      trajectory_execution_manager_ptr->ensureActiveControllersForJoints(std::vector<std::string>(1, "panda_joint1")));
```
Comment (rhaschke) [inline moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp:52]: Can't you just write:
```suggestion
      trajectory_execution_manager_ptr->ensureActiveControllersForJoints({"panda_joint1"});
```

File: moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp [hunk 3]
```diff
@@ -1,107 +1,103 @@
...
-
-  if (!tem.push(traj2))
-    ROS_ERROR("Fail!");
+TEST_F(MoveItCppTest, pushTrajectoryTest)
+{
+  ASSERT_TRUE(trajectory_execution_manager_ptr->push(traj1));
+  ASSERT_TRUE(trajectory_execution_manager_ptr->push(traj2));
+  traj1.multi_dof_joint_trajectory = traj2.multi_dof_joint_trajectory;
+  ASSERT_TRUE(trajectory_execution_manager_ptr->push(traj1));
+}
```
Comment (rhaschke) [inline moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp:74]: This is redundant to the next test. I suggest to drop it.

File: moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp [hunk 4]
```diff
@@ -1,107 +1,103 @@
...
+}  // namespace moveit_cpp
 
-  if (!tem.executeAndWait())
-    ROS_ERROR("Fail!");
+int main(int argc, char** argv)
+{
+  // if (ros::console::set_logger_level(ROSCONSOLE_DEFAULT_NAME, ros::console::levels::Debug))
+  // {
+  //   ros::console::notifyLoggerLevelsChanged();
+  // }
```
Comment (rhaschke) [inline moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp:93]: Please drop this. The debug level can be easily set via a `ROSCONSOLE_CONFIG_FILE`.

File: moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.test [hunk 5]
```diff
@@ -1,20 +1,37 @@
...
-  <!-- Load the URDF, SRDF and other .yaml configuration files on the param server -->
-  <include file="$(find moveit_resources)/pr2_moveit_config/launch/planning_context.launch">
-    <arg name="load_robot_description" value="true"/>
-  </include>
+    <!-- Specify the config files to use -->
+    <rosparam ns="test_execution_manager" command="load" file="$(find moveit_ros_planning)/trajectory_execution_manager/test/config/moveit_cpp.yaml" />
 
-  <!-- If needed, broadcast static tf for robot root -->
-  <node pkg="tf2_ros" type="static_transform_publisher" name="virtual_joint_broadcaster_1" args="0 0 0 0 0 0 world base_footprint" />
+    <!-- Planning Pipeline -->
```
Comment (rhaschke) [inline moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.test:5]: Have a look at [`test_environment.launch`](https://github.com/ros-planning/moveit_resources/blob/100bb2dfa8d92832f1b06f45bfc6e8cded26cb75/panda_moveit_config/launch/test_environment.launch#L1). This should launch all required stuff.

File: moveit_ros/planning/trajectory_execution_manager/src/trajectory_execution_manager.cpp [hunk 6]
```diff
@@ -1211,6 +1211,10 @@ void TrajectoryExecutionManager::execute(const ExecutionCompleteCallback& callba
 void TrajectoryExecutionManager::execute(const ExecutionCompleteCallback& callback,
                                          const PathSegmentCompleteCallback& part_callback, bool auto_clear)
 {
+  // skip
```
Comment (rhaschke) [inline moveit_ros/planning/trajectory_execution_manager/src/trajectory_execution_manager.cpp:1214]: Please add a comment why this is needed.

**Review by rhaschke — COMMENTED** [review]
Thanks for reactivating these tests. I have a few comments though:

File: moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp (hunk 2 above)
Comment (cambel) [reply moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp:52]: No, the ensureActiveCountrollersForJoints is only defined for a vector of strings. I could add an overload for just a string if needed

File: moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.test (hunk 5 above)
Comment (cambel) [reply moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.test:5]: I wrote the tests with `moveit_cpp` so the `test_environment.launch` does not work for that. I could create the configuration in the `panda_moveit_config` package instead.

File: moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp (hunk 1 above)
Comment (cambel) [reply moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp:1]: okay

File: moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp (hunk 3 above)
Comment (cambel) [reply moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp:74]: okay

File: moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp (hunk 4 above)
Comment (cambel) [reply moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp:93]: okay

File: moveit_ros/planning/trajectory_execution_manager/src/trajectory_execution_manager.cpp (hunk 6 above)
Comment (cambel) [reply moveit_ros/planning/trajectory_execution_manager/src/trajectory_execution_manager.cpp]: I added a more clear comment. The thing is that it crashes if no trajectory has been pushed.

File: moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp (hunk 2 above)
Comment (simonschmeisser) [reply moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp:86]: The angle brackets create an initializer list which should be implicitly converted to a vector

Comment (rhaschke) [reply moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp:86]: Simon is right, but he refers to the curly brackets actually :wink:
Please try the suggestion.

File: moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp [hunk 7]
```diff
@@ -32,76 +32,94 @@
...
-  traj2.joint_trajectory.points[0].positions.push_back(1.0);
-  traj2.multi_dof_joint_trajectory.joint_names.push_back("basej");
-  traj2.multi_dof_joint_trajectory.points.resize(1);
-  traj2.multi_dof_joint_trajectory.points[0].transforms.resize(1);
-
-  if (!tem.push(traj2))
-    ROS_ERROR("Fail!");
+  trajectory_execution_manager_ptr->execute();
+  auto last_execution_status = trajectory_execution_manager_ptr->waitForExecution();
+  ASSERT_TRUE(last_execution_status == moveit_controller_manager::ExecutionStatus::SUCCEEDED);
```
Comment (rhaschke) [inline moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp:99]: What about combining these two lines:
```suggestion
  ASSERT_EQUAL(trajectory_execution_manager_ptr->waitForExecution(), moveit_controller_manager::ExecutionStatus::SUCCEEDED);
```

File: moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp [hunk 8]
```diff
@@ -32,76 +32,94 @@
...
+TEST_F(MoveItCppTest, PushExecuteAndWaitTest)
+{
+  ASSERT_TRUE(trajectory_execution_manager_ptr->push(traj1));
+  ASSERT_TRUE(trajectory_execution_manager_ptr->push(traj2));
   traj1.multi_dof_joint_trajectory = traj2.multi_dof_joint_trajectory;
-  if (!tem.push(traj1))
-    ROS_ERROR("Fail!");
+  ASSERT_TRUE(trajectory_execution_manager_ptr->push(traj1));
+  auto last_execution_status = trajectory_execution_manager_ptr->executeAndWait();
+  ASSERT_TRUE(last_execution_status == moveit_controller_manager::ExecutionStatus::SUCCEEDED);
```
Comment (rhaschke) [inline moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp:109]: Same here:
```suggestion
  ASSERT_EQUAL(trajectory_execution_manager_ptr->executeAndWait(), moveit_controller_manager::ExecutionStatus::SUCCEEDED);
```

File: moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp [hunk 9]
```diff
@@ -32,76 +32,94 @@
...
+
+int main(int argc, char** argv)
+{
+  testing::InitGoogleTest(&argc, argv);
+  ros::init(argc, argv, "test_execution_manager");
 
-  if (!tem.executeAndWait())
-    ROS_ERROR("Fail!");
+  ros::AsyncSpinner spinner(4);
+  spinner.start();
```
Comment (rhaschke) [inline moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp:120]: Why do you need a spinner at all? There aren't any ROS messages to be handled by the test.

File: moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp [hunk 10]
```diff
@@ -32,76 +32,94 @@
...
-  if (!tem.ensureActiveControllersForJoints(std::vector<std::string>(1, "rj2")))
-    ROS_ERROR("Fail!");
+namespace moveit_cpp
+{
+class MoveItCppTest : public ::testing::Test
+{
+public:
+  void SetUp() override
+  {
+    nh_ = ros::NodeHandle("/test_execution_manager");
```
Comment (rhaschke) [inline moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp:58]: Why do you specify a namespace here? This isn't needed, is it?
```suggestion
```

File: moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.test [hunk 11]
```diff
@@ -1,20 +1,37 @@
-<?xml version="1.0"?>
 <launch>
-  <!-- Load the URDF, SRDF and other .yaml configuration files on the param server -->
-  <include file="$(find moveit_resources)/pr2_moveit_config/launch/planning_context.launch">
-    <arg name="load_robot_description" value="true"/>
-  </include>
+    <!-- Specify the config files to use -->
+    <rosparam ns="test_execution_manager" command="load" file="$(find moveit_ros_planning)/trajectory_execution_manager/test/config/moveit_cpp.yaml" />
```
Comment (rhaschke) [inline moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.test:3]: Dropping the namespace specification in the code above, you can drop it here (and a few other locations below) as well, which simplifies code:
```suggestion
    <rosparam command="load" file="$(find moveit_ros_planning)/trajectory_execution_manager/test/config/moveit_cpp.yaml" />
```

File: moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.test (hunk 5 above)
Comment (rhaschke) [reply moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.test:5]: You are right. You don't need a _move_group_ node when using MoveItCpp.
However, it doesn't harm either. Using `test_environment.launch` avoids all the boiler-plate code...

**Review by rhaschke — COMMENTED** [review]
Thanks for cleaning up. I have a few more comments to further streamline the code.

File: moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp (hunk 2 above)
Comment (cambel) [reply moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp:86]: Oh sorry, I didn't notice the curly brackets the first time. I thought you were referring to the string alone.

**Comment by cambel** [comment]
@rhaschke I simplified the configuration file using the `test_environment.launch` file.
The only thing I couldn't remove was the controllers' configuration. I know that `demo.launch` is loading the controllers' configuration to the namespace `move_group` but the test can't seem to find them if they are not under the test namespace which is the test-name itself `test_execution_manager`.
Do you have any idea why?

File: moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp (hunk 7 above)
Comment (cambel) [reply moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp]: okay

File: moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp (hunk 8 above)
Comment (cambel) [reply moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp]: okay

File: moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp (hunk 9 above)
Comment (cambel) [reply moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp]: okay

File: moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp (hunk 10 above)
Comment (cambel) [reply moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp]: I remove the namespace but the NodeHandle seems necessary for instantiating MoveItcpp

File: moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.test (hunk 11 above)
Comment (cambel) [reply moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.test]: okay

File: moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp (hunk 10 above)
Comment (rhaschke) [reply moveit_ros/planning/trajectory_execution_manager/test/test_execution_manager.cpp]: Sure.

**Comment by rhaschke** [comment]
> The test can't seem to find the controller params if they are not under the test namespace.
> Do you have any idea why?

As you explained yourself: The parameters are looked up in the local namespace of the node. The move_group's params are not considered.

**Review by rhaschke — APPROVED** [review]

**Pull request merged**